scripts/generate-changelog

```diff
@@ -0,0 +1,328 @@
+      pageInfo {
+        hasNextPage
+        endCursor
+      }
+    }
+  }
+}
+"""
+
+
+def process_changelog(resource_type: str, changelog: dict, releases: set, projects: set, no_project_prs: list):
+    if resource_type == PULL_REQUESTS:
+        items = process_pr(releases)
+        time_field = 'mergedAt'
+    elif resource_type == ISSUES:
+        items = process_issue()
+        time_field = 'closedAt'
+    else:
+        print(f"[process_changelog] Invalid type: {resource_type}")
+        sys.exit(1)
+
+    for item in items:
+        if len(item['projectCards']['nodes']) == 0:
+            if resource_type == PULL_REQUESTS:
+                no_project_prs.append(item)
+            continue
+
+        project = item['projectCards']['nodes'][0]['project']['name']
+        if not release_project(project, projects):
+            continue
+
+        if project not in changelog:
+            changelog[project] = {
+                FEATURES: [],
+                PERFORMANCE: [],
+                BUGS_FIXED: [],
+                PRS: [],
+            }
+
+        labels = set()
+        for label in item['labels']['nodes']:
+            labels.add(label['name'])
+        category = rules(labels)
+        if resource_type == ISSUES and category == INVALID:
+            continue
+        if resource_type == PULL_REQUESTS:
+            if '[bot]' in item['title']:  # skip auto-gen PR, created by our github actions workflows
+                continue
+            category = PRS
+
+        changelog[project][category].append({
+            "number": item['number'],
+            "title": item['title'],
+            "url": item['url'],
+            "time": item[time_field],
+        })
+
+
+def process_pr(releases: set):
+    pr = []
+    for rel in releases:
+        pr.extend(fetch(PULL_REQUESTS, {'baseRefName': f"branch-{rel}"}))
+    return pr
+
+
+def process_issue():
+    return fetch(ISSUES)
+
+
+def fetch(resource_type: str, variables: dict = None):
+    items = []
+    if resource_type == PULL_REQUESTS and variables:
+        q = query_pr
+    elif resource_type == ISSUES:
+        q = query_issue
+        variables = {}
+    else:
+        return items
+
+    has_next = True
+    while has_next:
+        res = post(q, variables)
+        if res.status_code == 200:
+            d = res.json()
+            has_next = d['data']['repository'][resource_type]["pageInfo"]["hasNextPage"]
+            variables['after'] = d['data']['repository'][resource_type]["pageInfo"]["endCursor"]
+            items.extend(d['data']['repository'][resource_type]['nodes'])
+        else:
+            raise Exception("Query failed to run by returning code of {}. {}".format(res.status_code, q))
+    return items
+
+
+def post(query: str, variable: dict):
+    return requests.post('https://api.github.com/graphql',
+                         json={'query': query, 'variables': variable},
+                         headers={"Authorization": f"token {GITHUB_TOKEN}"})
```

Comment:
Now that `GITHUB_TOKEN` is no longer a global, how is this method looking up `GITHUB_TOKEN`?  Doesn't it need to be passed as an argument?

Reply:
python takes variables under `__name__ == '__main__'` as global, I could just make it as part of args